pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,487 @@
+	numAllowed := MaxConcurrentRaftTraces.Get(&r.st.SV)
+	numRegisteredReplica := r.numRegisteredReplica.Load()
+
+	// This can only occur if the numAllowed setting has changed since a
+	// previous call to reserveSpace. If this happens flush all our current
+	// traces and don't register this request. Note that when this happens we
+	// also wont't log this request.
+	if numRegisteredReplica > numAllowed {
+		log.Infof(r.ctx, "flushing all traces due to setting change")
+		r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+			r.removeEntry(index)
+			return true
+		})
+		return false
+	}
+
+	if numAllowed == 0 {
+		return false
+	}
+
+	// The maximum number of traces has been reached for the store. We don't
+	// register tracing and return false.
+	numRegisteredStore := r.numRegisteredStore.Load()
+	if numRegisteredStore >= numAllowed {
+		return false
+	}
+
+	// Only increment the number of registered traces if the numRegistered
+	// hasn't changed. In the case of an ABA update, it does not break the
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available. We will not register this trace if
+	// someone else is concurrently registering a trace on this store, but this
+	// is acceptable as it is a rare case.
+	registerSucceeded := r.numRegisteredStore.CompareAndSwap(numRegisteredStore, numRegisteredStore+1)
+	if registerSucceeded {
+		// Add one unconditionally to the replica count.
+		r.numRegisteredReplica.Add(1)
+	}
+	// Note we can't assert numRegisteredStore <= numAllowed because if the
+	// setting is changed it can be temporarily violated on other replicas.
+	return registerSucceeded
+}
+
+// tryStore attempts to store this value. If the index is already in the map it
+// will not store this entry and return false. It will also decrement counters
+// that were incremented by reserveSpace.
+// This is a rare case where we already have the index in the map. We
+// don't want to store this entry, but also need to decrement the
+// counter to avoid double tracing.
+func (r *RaftTracer) tryStore(tv *traceValue) *traceValue {
+	if existingTv, loaded := r.m.LoadOrStore(tv.traced.Index, tv); loaded {
+		r.numRegisteredReplica.Add(-1)
+		r.numRegisteredStore.Add(-1)
+		if tv.mu.propSpan != nil {
+			tv.logf(2, "duplicate registration ignored - existing trace: %s", existingTv)
+			tv.mu.propSpan.Finish()
+			// Don't allow additional tracing to this context.
+			tv.mu.propCtx = nil
+			tv.mu.propSpan = nil
+		}
+		existingTv.logf(2, "additional registration for same index: %s", tv)
+		return existingTv
+	}
+	return tv
+}
+
+// storeEntry stores the TracedEntry without a tracing context and returns the
+// newly created traceValue.
+func (r *RaftTracer) storeEntry(te kvserverpb.TracedEntry) *traceValue {
+	tv := traceValue{traced: te}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+	tv.mu.seenMsgAppResp = make(map[raftpb.PeerID]bool)
+	return r.tryStore(&tv)
```

Comment:
Perhaps it would be slightly cleaner to remove the `tryStore` and do the logic here:

```golang
existing, loaded := r.m.LoadOrStore(te.Index, tv)
if loaded {
	r.destroy(tv) // destroys the context and decrements counters
	return existing, false
}
return tv, true
```

And reuse the `destroy` func in `removeEntry` which does the same thing as the "destroy" part of the `tryStore`.

Reply:
Done.